platform-api/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/AttributeFilter.java

```diff
@@ -0,0 +1,196 @@
+                                                                                             ValueStorageException {
+        Map<String, List<String>> runtimeAttributes = new HashMap<>();
+
+        for (Attribute attribute : projectType.getAttributes()) {
+            if (attribute.isVariable()) {
+                Variable variable = (Variable)attribute;
+                ValueProviderFactory factory = variable.getValueProviderFactory();
+
+                List<String> value;
+                if (factory != null) {
+                    value = factory.newInstance(projectFolder).getValues(variable.getName());
+
+                    if (value == null) {
+                        throw new ProjectTypeConstraintException("Value Provider must not produce NULL value of variable " +
+                                                                 variable.getId());
+                    }
+                } else {
+                    value = projectConfig.getAttributes().get(attribute.getName());
+                }
+
+                if ((value == null || value.isEmpty()) && variable.isRequired()) {
+                    throw new ProjectTypeConstraintException("No Value for provider defined for required variable " + variable.getId());
+                } else {
+                    runtimeAttributes.put(variable.getName(), value);
+                }
+            }
+        }
+
+        return runtimeAttributes;
+    }
+
+    /**
+     * Adds runtime attributes to project. Runtime are attributes which are calculated in runtime during getting or updating project.
+     * The method analyzes attributes from project and internal modules recursive.
+     *
+     * @param projectConfig
+     *         project configuration which contains information about project and internal modules
+     * @param projectFolder
+     *         folder in which project or module is located
+     * @throws ServerException
+     *         error occurs during we get project from workspace
+     * @throws ForbiddenException
+     *         if user which perform operation doesn't have access to the requested project
+     * @throws ProjectTypeConstraintException
+     *         if we create project types without required parameters
+     * @throws ValueStorageException
+     *         when we have exception during getting values from value factory in attribute
+     */
+    public void addRuntimeAttributesToProject(ProjectConfigDto projectConfig,
+                                              FolderEntry projectFolder) throws ProjectTypeConstraintException,
+                                                                                ForbiddenException,
+                                                                                ValueStorageException,
+                                                                                ServerException {
+        addAttributesToProject(projectConfig, projectFolder, RUNTIME);
+    }
+
+    enum AttributeType {
```

Comment:
Yes. I use this enum only in this class for convenience and to do code more readable.